receiver/nats.go

```diff
@@ -0,0 +1,159 @@
+	UserCreds     string            `doc:"Nats credentials file path"`
+	NKeyFile      string            `doc:"Nats nkey file path"`
+	Insecure      bool              `doc:"TLS InsecureSkipVerify"`
+	conOpts       *[]nats.Option
+	natsCon       *nats.Conn
+	wg            sync.WaitGroup
+}
+
+// Verify configuration
+func (n *Nats) Verify() error {
+	if n.Handler.Name == "" {
+		return skogul.MissingArgument("Handler")
+	}
+	if n.Subject == "" {
+		return skogul.MissingArgument("Subject")
+	}
+	if n.Servers == "" {
+		return skogul.MissingArgument("Servers")
+	}
+	//User Credentials
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		//Cred file contains nkey.
+		return fmt.Errorf("Please configure usercreds or nkeyfile.")
+	}
+
+	return nil
+}
+
+func (n *Nats) Start() error {
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.conOpts = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.UserCreds != "" {
+		*n.conOpts = append(*n.conOpts, nats.UserCredentials(n.UserCreds))
+	}
+
+	//Plain text passwords
+	if n.Username != "" && n.Password != "" {
+		if n.TLSClientKey != "" {
+			natsLog.Warnf("Using plain text password over a non encrypted transport!")
+		}
+		*n.conOpts = append(*n.conOpts, nats.UserInfo(n.Username, n.Password))
+	}
+	//TLS authentication, Note: Fix selfsigned certificates.
+	if n.TLSClientKey != "" && n.TLSClientCert != "" {
+		cert, err := tls.LoadX509KeyPair(n.TLSClientCert, n.TLSClientKey)
+		if err != nil {
+			return fmt.Errorf("error parsing X509 certificate/key pair: %v", err)
+		}
+
+		cp, err := skogul.GetCertPool(n.TLSCACert)
+		if err != nil {
+			return fmt.Errorf("Failed to initialize root CA pool")
+		}
+
+		config := &tls.Config{
+			InsecureSkipVerify: n.Insecure,
+			Certificates:       []tls.Certificate{cert},
+			RootCAs:            cp,
+		}
+		*n.conOpts = append(*n.conOpts, nats.Secure(config))
+	}
+
+	//NKey auth
+	if n.NKeyFile != "" {
+		opt, err := nats.NkeyOptionFromSeed(n.NKeyFile)
+		if err != nil {
+			natsLog.Fatal(err)
+		}
+		*n.conOpts = append(*n.conOpts, opt)
+	}
+
+	//Log disconnects
+	*n.conOpts = append(*n.conOpts, nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
+		natsLog.WithError(err).Error("Got disconnected!")
+	}))
+	//Log reconnects
+	*n.conOpts = append(*n.conOpts, nats.ReconnectHandler(func(nc *nats.Conn) {
+		natsLog.Info("Reconnected")
+	}))
+	//Always try to reconnect
+	*n.conOpts = append(*n.conOpts, nats.RetryOnFailedConnect(true))
+	//Try to reconnect forever
+	*n.conOpts = append(*n.conOpts, nats.MaxReconnects(-1))
+
+	var err error
+	n.natsCon, err = nats.Connect(n.Servers, *n.conOpts...)
+	cb := func(msg *nats.Msg) {
+		natsLog.Debugf("Received message on %v", msg.Subject)
+		if err := n.Handler.H.Handle(msg.Data); err != nil {
+			natsLog.WithError(err).Warn("Unable to handle Nats message")
+		}
+		return
+	}
+
+	n.wg.Add(1)
+	if err != nil {
+		natsLog.Errorf("Encountered an error while connecting to Nats: %v", err)
+	}
```

Comment:
I think this err-check should live closer to the place it was generated; right after `nats.Connect(...)`. The callback function in-between here can be defined elsewhere so that it is "not in the way" (e.g. above the connect).